### ISSUE TITLE
Rule to ensure explicit paths as required by ES Modules

### ISSUE THREAD

**Issue by gustavopch** [import-js/eslint-plugin-import#2319]
I've tried to find some way to ensure explicit paths as required by ES Modules, but couldn't find anything. To be clear, I want a rule that forces the code to use explicit extensions and indexes.

```js
import x from './some-file' // ❌ Implicit extension
import x from './some-directory' // ❌ Implicit index

import x from './some-file.js' // ✅ Explicit extension
import x from './some-directory/index.js' // ✅ Explicit index
```

Is there a rule already for that?

**Comment by Haroenv** [comment]
this would be import/extensions if I remember it correctly, but note that it doesn't work yet with typescript #2011, #2111, #2170 (not sure yet what the solution is)

**Comment by ljharb** [comment]
Handling this properly would likely require “exports” support, and native ESM support, which we (and much of the ecosystem) don’t have yet. This plugin still assumes you’re not using native ESM in node, and are transpiling with Babel or typescript.

**Comment by ljharb** [comment]
@wojtekmaj once `resolve` supports "exports", and then ESM, then this plugin will, and it will be revisited.

However, it remains the case that ESM-only packages fail to get adoption because they aren't usable in CJS projects, and CJS-only (or dual) packages are the easiest to use and adopt, so the ecosystem has pretty clearly demonstrated it's a low priority.

**Comment by satazor** [comment]
I wanted the exact same feature, so I created a small plugin: https://github.com/satazor/eslint-plugin-import-esm. Hope it helps until this plugin gets the feature requested in this ticket.

Please note however, it doesn’t support `exports` in the package.json.
